### ISSUE TITLE
Support Ping requests to report a better 'time on page'

### ISSUE THREAD

**Issue by mattab** [piwik/piwik#2041]
Currently, the last page view is set with a 'time on page' of zero seconds. 

Instead, we could regularly ping the Piwik server with a request that notifies Piwik that the user is still on the website.

This request should not record new page view, only increase the visit_total_time counter.

Also, we should only throw this request when the user is actually looking at the page, not doing something else in some other tabs. 

The risk with this feature is that instead of improving the report accuracy, we end up inflating numbers: a lot of users leave pages open in other tabs for hours and don't do anything about it. We should heavily test to ensure we never throw a request on a page where the user didn't interact with recently. Some interactions we can maybe detect: mouse click, mouse move, etc.

Of course, we should not make everything slow by adding handlers on mouse move, etc.

Also, once this feature is implemented, a new checkbox should be added to the Advanced section in the 'Tracking Code' admin page. Checking it will generate JS code that enables this feature.

**Comment by IPL01** [comment]
Has this feature been implemented now? 

I get a lot of single page views and they all report back as 0s! NOT GOOD and ultimately paints an unhelpful analytical report. 

I have read through all the other forum posts about why this is the case and why things like a setHeartBeatTimer are difficult to implement... 

But this was all a really long time ago. Please advise whether there are any new fixes/mods to improve on this.

Many thanks.

**Comment by Glisse1** [comment]
Yay, short-term milestone! +1 :)

**Comment by mattab** [comment]
moved into 2.14.0 milestone, as this is such a useful feature, let's get it done :+1:

**Comment by hpvd** [comment]
as already referenced above (#5208) taking existing events into account should help detect if page is only still opened in background tab without a performance problem.
Another thing do we need a "timeout" option e.g. no one looks a website longer than 15min to keep statistics fine? or 5min after last event?

**Comment by hpvd** [comment]
since the function
"active live on open page/no live on page"
could be not only be used for "better time on page for last page" (goal of this ticket)

but could also greatly be used as function within other features like
- "detect page is only opend in background"
(lots of browser tabs opend, even afetr browser restart..)
- "do not count automatic visits from website watchers" #3556 and #5845
- ...

*=> what do you think to make it a general function within Piwik?*

It could bundle all technologies to decide
"active live on open page / no live on page"
- Via triggered events
- via additional mouse movement tracking
- via additional scroll tracking
- via ping
- via timeout
and simply made the result available for every feature which needs it?

=> should we open a new ticket for this?

**Comment by masteranalyze** [comment]
I think the best solution ,is to use : Advanced Content Tracking like with Universal Analytics.
For reference i think this is the best method,that can simple solve the problem,and be quickly added on next piwik release:
http://cutroni.com/blog/2014/02/12/advanced-content-tracking-with-universal-analytics/
Original article: http://cutroni.com/blog/2012/02/21/advanced-content-tracking-with-google-analytics-part-1/

here’s what we’re going to track:

    How many people scroll
    When a person starts to scroll
    When a person reaches the end of an article (not the end of the page, but the end of the article or post area)
    When a person reaches the botton of the page (the bottom of the HTML)
    Which website visitors are scanning my articles and which are reading my articles
Think about the value here! We will be able to get an accurate measure of which articles are actually read. We can even see which articles are so engaging that visitors continue through the comments to the botto of the page. Very useful stuff.

Tracking Technique

All of the above can be tracked with Event Tracking. The concept is that we will fire events when certain actions happen. Specifically we’re going to fire events based on visitors scrolling down the page.

Finally :+1: 
 Increase in the Average Session Duration. Events also increase the Time on Page measurement on exit pages. This may cause an increase in the Average Session Duration metric.

Increasing Bounce Rate is not Necessarily a Bad Thing!

If a visitor comes to you site, reads a post in-depth, and then leaves, should they really be classified as a bounce? Probably not. With this implementation, your bounce-rate will provide a more accurate measure of engagement and it will solve the real time spend on the website!

I hardly wait for this implemenation to be alive!

**Comment by masteranalyze** [comment]
Except this events,in order to even have an more accurate time it can be added:

1)Track user that print pages : 
Here is the code for google analytics  that can be easily converted to piwik to track this events:
https://www.savio.no/analytics/how-to-track-printed-pages-in-google-analytics

2)Track Downloads & Other External Links,track mail to,telephone clicks.
Here is the code for google analytics  that can be easily converted to piwik to track this events:
http://www.blastam.com/blog/index.php/2013/09/howto-track-downloads-links-universalanalytics

3)Social actions on page engagement:
Here is the code for google analytics  that can be easily converted to piwik to track this events:
http://www.optimizesmart.com/social-interactions-tracking-through-google-analytics/

It`s working for majors:Twitter, Facebook, Google Plus & Linkedin,others can be added as need it.Maybe pininterest,stumble,etc.

If we track the 5 events from the cutroni example+3 that adds up here like :print,download,external links,and social interaction,we will be able to get an more better ideea of the interaction on that page,and not only the time will be more accurate,it will be the bounce rate too more accurate,as that user has done action and interacted with that webpage,so it should not be consider it an bounce as the user interacted with that webpage by sharing,mail to,phone,read,etc.

I think this can be implemeneted right the way with the help of the events from piwik.

**Comment by mattab** [comment]
Not only have many people commented on this issue but in the forums this ticket was linked to about [40 times](http://forum.piwik.org/search.php?15,search=2041,author=,page=1,match_type=ALL,match_dates=0,match_forum=ALL,match_threads=0) in the last few years.

**Comment by Glisse1** [comment]
Just wanted to share the forum post from a year ago with suggestion as of what direction the visual displaying of real-time status (active, idle, reading, writing, etc) could be going...  knowing the true active time on page in statistics is important, but I believe it`s just as important is to have some sort of visual representation in real-time widget

http://forum.piwik.org/read.php?3,112235,page=1#msg-112235

**Comment by masteranalyze** [comment]
"Also, once this feature is implemented, a new checkbox should be added to the Advanced section in the 'Tracking Code' admin page. Checking it will generate JS code that enables this feature. "

Exactly ,this is how it should be for ping in,and for the events method,so users can simple enable from events,if they wish to enable : bounce handle,social network sharing,print,download,etc.
and once enabled,the code to automatically track them,to not have like now,to add alot of scripts for those events.
For example this:

script
setTimeout("_paq.push(['trackEvent', 'time spent', '>15sec', 'good reading'])",15000); 
/script 
This will make more simple to enable events and users on page time will improve and we will know it real,and also the bounce rate will be adjusted.
By solving this we will solve not only the 1st page timer ,but also the real bouncing rate,as from my point of view if an user spend 15-30 seconds or even more on just that webpage and he is interacting:sharing,printing,downloading or any action that can be done into an webpage,that is not an bounce at all.
That means the user find very usefully that blog post or whatever,and we cannot now that,unless this feature will be ready on 2.14.
Hopefully it will be alive!

**Comment by mattab** [comment]
Summary of discussion so far:
* piwik.js:  `sendRequest` mechanism can be reused
* need to decide on "ping" strategy eg. ping first time after 20s, then ping less and less often eg. [20, 40, 60, 100, 160, 260, ....]  ?
* we need to ping only when the page is actively viewed (we need to detect active tab/window somehow, maybe Page Visibility API can do this?)
* define & document a new Tracking API parameter, eg `ping=1`

Maybe you have more thoughts, cc @diosmosis

**Comment by hpvd** [comment]
thanks for the summary. Hopefully it's ok when I would like to ask to clarify the target even more:

1) will this feature be an additional thing to other already available "site is active" signals like events (see user posts of robocoder, masteranalyze and me) so it mainly brings new active window detection?
And Piwik host will combine this and notice which signal came "last"?

2) or will it use these available signals to determine site is active, add active window detection and send a ping if there is any activity?

3) or nothing of these ;-)

Many thanks!

**Comment by mattab** [comment]
@hpvd currently there is no feature - the code related to heartbeat timer in piwik.js is actually unused we could/should remove it as well!

**Comment by masteranalyze** [comment]
Hey guys,
Matt,hpvd,i think the best method is 1 " will this feature be an additional thing to other already available "site is active" signals like events (see user posts of robocoder, masteranalyze and me) so it mainly brings new active window detection?
And Piwik host will combine this and notice which signal came "last"? "

This is how it should be .
That way based on events,and user actions,we will have accurate TIMING+Bouncing rates!

About the : need to decide on "ping" strategy eg. ping first time after 20s, then ping less and less often eg. [20, 40, 60, 100, 160, 260, ....] ? ,ping strategy,i think from 15seconds ,to 15 seconds is okay or even from 20 seconds to 20 seconds,but this also can depends on user shared enverimonths,if the user haves Shared hosting,and this ping in eats lots of resources from the server,cpu,ram,i think their should be an recomandation for shared ,vps and dedicated server.
An dedicated server can have no problem to handle from 15 seconds to 15 seconds,but an shared server might have an problem,and maybe is good to put it from 30 or 60 seconds .
Anyhow an minimum value should be 15,and not less,and we let the user to decide.


we need to ping only when the page is actively viewed (we need to detect active tab/window somehow, maybe Page Visibility API can do this?)

Exactly,we need to ping in,only when the page is actively viewed,and this can be done with the events,that we talked more earlier.If they are no events,basically the page is maybe left in the tab for hours,and user is watching an youtube clip,so it`s not in your page tab,so basically this will just inflate the numbers witch is not good.

The events should also be there,and easily accesibile,as options in piwik,and we let the user freedom to decide,maybe some of us only needs some of the events,or all,it depends from case to case.

@matt : define & document a new Tracking API parameter, eg ping=1 (or better name!)

I think tracking api parameter, ping =1 ,is an good name,but what 1 will represent ?
1 will represent that PING is actively ?Or it will ping from 1 second to another ?
Basically 1 should Active Ping In,and 0 - Disabled Ping IN.

Hopefully we will see this in the new piwik 2.14.

**Comment by hpvd** [comment]
"1)additional thing to other already available "site is active" signals"
that's also imho, and was the reason for creating https://github.com/piwik/piwik/issues/7607

**Comment by masteranalyze** [comment]
so this will work,basically on 15 seconds to 15 seconds to the end user pings?
or via events triggers ?

I agree,hpvd,there is no great "Event tracking" available in piwik.

**Comment by hpvd** [comment]
@masteranalyze now standard event tracking is fine for me. What I was describing was the time the ticket was started...in 2011 ther was no event tracking at all: so you couldn't determine at all if there is some on the site. If you are using event tracking today, you got lots of activity signal from each site during the usage by the visiter

**Comment by masteranalyze** [comment]
@hpvd  is clearly that from that time to today,piwik did have an great evolution,but i think it can be even more great,as except the fact the anaytics are private,there is one more good thing and an big advantage for PIWIK - Server Side Tracking,witch is not possible neither in Google Analytics,neither in StatCounter,neither in Clicky or any other stats available on the market today.

The event tracking we have today,of course is more good,that what it was missing in 2011,i totally agree with you,but nowdays for every event tracking,you need to add one more line code into the javascript,one more for another event and so on,it will be alot more easily,if in the event tracking back -end it will be events with options : YES ,NO ,and addiacently the options to it,so everybody can Enable or use them,right now Events are not quite so easily setup for someone who does not write code.
This is for example for : ADJUSTED BOUNCED RATE,but if you don`t add it to your code of tracking,it won`t track this event,and mostly your bounce rate will be higher,and time spend on the page less.
script
setTimeout("_paq.push(['trackEvent', 'time spent', '>15sec', 'good reading'])",15000);
/script 
Maybe we can make an list of all this,scripts for events how they should be,and further on,they can be implemented and be available from backend Event Tracking!

"If you are using event tracking today, you got lots of activity signals from each site during the usage by the visitor." - i totally agree with you,with the pointing of that only people who setup the events and puts additionally script,can track the events,or if the software your using,haves integer piwik,you will have those events,else not.

" So the ping only needs to be an additional possibility to determine if there is activity.
But of course the signals from the events first have to be included in the determination of active/passive..."      - like i said and i agree with you,that this ping in,normaly should be an option from backend,that everybody to be able to setup up,for example on 15 seconds,on 30seconds,to be the ping in .
And we also do needs all the Events,in Back END,to check the user activity,if he is active or passive.
In administration of piwik,we don`t have Events ,to be easily setup,with yes or no,and to not have to add,more and more code,into the piwik original tracking code.

hpvd,create an new ticket for PIWIK EVENTS SCRIPTS,and there we can put an list of EVENTS,and the accordingly script,but afterwards,how we manage to put them into the backend of piwik?!
What will be the tracking script for this action : Social actions on page engagement:
Here is the code for google analytics that can be easily converted to piwik to track this events:
http://www.optimizesmart.com/social-interactions-tracking-through-google-analytics/ ?!
I think if we all contribute and make an list,afterwards,that list can be complected along the time!

**Comment by hpvd** [comment]
@masteranalyze
a ticket for collecting standard events and the code snippets belonging to them
is a great idea for a future improvement - please start the ticket!!

**Comment by masteranalyze** [comment]
@hpvd i opened this https://github.com/piwik/piwik/issues/8141 ,for posting colection of standard events and js code.
The thing is,how we,do implement them in piwik back-end to be easily accesibile?Even for non-coders.

**Comment by diosmosis** [comment]
Fixed in #8069.

**Comment by robocoder** [comment]
@diosmosis \o/

**Comment by jgebal** [comment]
Hi all,
+1 to the entire idea.
I'm using Piwik for my blog and I continuously see high bounce rates and no time on page for single page views.
And a short suggestion.
The feature is complex so don't try to make it perfect. Just do something.
If the discussion takes 4,5 years, there is something terribly wrong.
Make it Lean, use inspect and adapt. Release it with option to disable/enable the feature and people will give you feedback how it works what works what does not work.
Don't try to figure it all by yourself :)
The lack of any kind of real metric on single page views is really annoying.
Cheers and thanks for all the great job done on Piwik.

**Comment by ReessKennedy** [comment]
+1 to this as well. And agreed with @jgebal. 

This would be really great option to enable!

**Comment by mattab** [comment]
@ReessKennedy feel free to post in the other issue: https://github.com/piwik/piwik/issues/8225 (this issue here is closed)

**Comment by mattab** [comment]
Feature is now documented at [Accurately measure the time spent on each page](http://developer.piwik.org/guides/tracking-javascript-guide#accurately-measure-the-time-spent-on-each-page)

**Comment by hpvd** [comment]
many thanks for documenting this!
Is there a setting how often website send this alive information?
not every body needs it every second, but every 60seconds maybe to long for others...
and of course you may need to find an individual balance between accuracy and  server-load

**Comment by VincentSC** [comment]
This is documented in the changelog: http://developer.piwik.org/changelog

On 18/12/15 11:43, hpvd wrote:
>
> many thanks for documenting this!
> Is there a setting how often website send this alive information?
> not every body needs it every second, but every 60seconds maybe to 
> long for others...
> and of course you may need to find an individual balance between 
> accuracy and server-load
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/piwik/piwik/issues/2041#issuecomment-165744819>.
>

-- 
___________________________________________________________
  V.G.Hindriksen MSc., CEO
  StreamComputing BV - http://www.streamcomputing.eu
  Location: Amsterdam, Netherlands, Europe
  Cell: +31 6 45400456
NEW PHONE NUMBER: +31 854865760
  KvK-number (Chamber of Commerce): 61901873
  Rabo Bank (IBAN/SEPA): NL89RABO0113885571 - BIC: RABONL2U
  OpenCL Newsletter: http://bit.ly/OpenCLnewsletter
  OpenCL Twitter   : http://twitter.com/StreamComputing
  OpenCL Blog      : http://www.streamcomputing.eu/blog
  PGP Public key   : http://bit.ly/VincentHindriksenPGP
___________________________________________________________

**Comment by hpvd** [comment]
thanks for pointing there.
*To use this feature use tracker.enableHeartBeatTimer(); or _paq.push(['enableHeartBeatTimer']);. By default, a ping request will be sent every 15 seconds. You can specify a custom ping delay (in seconds) by passing an argument, eg, tracker.enableHeartBeatTimer(10); or _paq.push(['enableHeartBeatTimer', 10]);.*
If this is working fine it should be added to docu as well..

**Comment by mattab** [comment]
@hpvd improved the doc, and also created New FAQ: [How do I improve accuracy of “Time spent on page” and “Visit duration” metrics?](https://piwik.org/faq/how-to/faq_21158/)

**Comment by hpvd** [comment]
Thanks very much!

**Comment by masteranalyze** [comment]
good work @mattab and the team!
